libnmstate/route_rule.py

```diff
@@ -193,27 +193,53 @@ def gen_metadata(self, route_state, ifaces):
         """
         route_rule_metadata = {}
         for route_table, rules in self._rules.items():
-            iface_name = self._iface_for_route_table(
-                route_state, route_table, ifaces
-            )
-            route_rule_metadata[iface_name] = {
-                Interface.IPV4: [],
-                Interface.IPV6: [],
-            }
-            if rules != self._cur_rules[route_table]:
-                route_rule_metadata[iface_name][
-                    BaseIface.RULE_CHANGED_METADATA
-                ] = True
+            rules_ipfamily = {Interface.IPV4: set(), Interface.IPV6: set()}
+            cur_rules_ipfamily = {Interface.IPV4: set(), Interface.IPV6: set()}
+            for rule in self._cur_rules[route_table]:
+                cur_rules_ipfamily[Interface.IPV6].add(
+                    rule
+                ) if rule.is_ipv6 else cur_rules_ipfamily[Interface.IPV4].add(
+                    rule
+                )
             for rule in rules:
-                family = Interface.IPV6 if rule.is_ipv6 else Interface.IPV4
-                route_rule_metadata[iface_name][family].append(rule.to_dict())
+                rules_ipfamily[Interface.IPV6].add(
+                    rule
+                ) if rule.is_ipv6 else rules_ipfamily[Interface.IPV4].add(rule)
+            for ip_family in (Interface.IPV4, Interface.IPV6):
+                if len(rules_ipfamily[ip_family]) == 0:
+                    continue
+                iface_name = self._iface_for_route_table(
+                    route_state, route_table, ifaces, ip_family
+                )
+                if route_rule_metadata.get(iface_name) is None:
+                    route_rule_metadata[iface_name] = {
+                        Interface.IPV4: [],
+                        Interface.IPV6: [],
+                    }
+                if rules_ipfamily[ip_family] != cur_rules_ipfamily[ip_family]:
+                    route_rule_metadata[iface_name][
+                        BaseIface.RULE_CHANGED_METADATA
+                    ] = True
+                for rule in rules_ipfamily[ip_family]:
```

Comment:
These nested `for` loop will be a problem when facing large amount of route rules. Please improve the performance of it.